lib/elixir/test/elixir/kernel/impl_test.exs

```diff
@@ -28,6 +28,31 @@ defmodule Kernel.ImplTest do
     @callback foo(any, any, any) :: any
   end
 
+  defmodule UseBehaviourWithoutImpl do
+    @callback foo() :: any
+    @callback bar() :: any
+
+    defmacro __using__(_opts) do
+      quote do
+        @behaviour Kernel.ImplTest.UseBehaviourWithoutImpl
+        def foo(), do: :auto_generated
+      end
+    end
+  end
+
+  defmodule UseBehaviourWithImpl do
+    @callback foo() :: any
+    @callback bar() :: any
+
+    defmacro __using__(_opts) do
+      quote do
+        @behaviour Kernel.ImplTest.UseBehaviourWithImpl
+        @impl true
+        def foo(), do: :auto_generated
```

Comment:
The question I would have is: If the behaviour has three callbacks and we provide two default implementations but we set `@impl true` to only one of them... would that be checked?

Reply:
Also what would happen if we declare the `@impl` for a non-existing callback? It seems that now we are simply skipping the checks. I think we want to keep the code mostly as is today... we just want to avoid warnings on `warn_missing_impls`.

I think this would be a heuristic. If there is a missing implementation, we get the context for the missing implementation (maybe nil). We only run the missing implementation check for other functions that have exactly the same context.

I think this would actually allow us to simplify the overridable logic we have today but we can do that as another PR.